Introduce Internal Data Transform RPCs

rockwotj opened the pull request:
This introduces the internal RPCs needed for producing the output of Wasm powered data transforms.

The justifications for using internal RPCs instead of the Kafka API:
- Our internal RPC client is not designed to be low latency or high throughput
- Using internal RPCs allows us to use a seperate set of accounting, limits, quotas and metrics specific to transforms
- Internal RPCs allows us to re-use internal authentication mechanism (mTLS, etc) instead of needing to mess with ephemeral credentials

## Backports Required

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v23.2.x
- [ ] v23.1.x
- [ ] v22.3.x

## Release Notes

* none

michael-redpanda commented on src/v/transform/rpc/serde.cc:
```diff
@@ -0,0 +1,48 @@
+transformed_topic_data::transformed_topic_data(
+  model::topic_partition tp, model::record_batch b)
+  : tp(std::move(tp)) {
+    ss::chunked_fifo<model::record_batch> batches;
+    batches.reserve(1);
+    batches.push_back(std::move(b));
```
question: Did you mean to declare a new instance of `batches` here?

rockwotj replied:
Nope :)

michael-redpanda commented on src/v/transform/rpc/serde.cc:
```diff
@@ -0,0 +1,48 @@
+  : tp(std::move(tp))
+  , batches(std::move(b)) {}
+
+transformed_topic_data transformed_topic_data::share() {
+    ss::chunked_fifo<model::record_batch> shared;
+    shared.reserve(batches.size());
+    for (auto& batch : batches) {
+        shared.push_back(batch.share());
+    }
+    return {tp, std::move(shared)};
+}
```
question: Maybe this is required by serde, but why not use a copy constructor or a copy assignment operator

rockwotj replied:
The copy constructor/assignment is disabled for batches. And this shares the underlying iobuf (no copies), it can be useful if you want to keep an iobuf :)

rockwotj replied:
For example if you wanted to do retries, you'd just share a batch each time and the underlying iobuf would be ref counted and not copied as far as I understand it.

michael-redpanda reviewed (commented):
🔥 

This looks great, couple of things:

question: Any reason this has to live in transform?  Can it be usable by other 'services' (e.g.  a service that may need to produce a log to a topic)

nit: Commit messages don't follow the style defined in [CONTRIBUTING.md]( https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md). First line should be `area[/detail]: short desc`

michael-redpanda commented on src/v/transform/rpc/client.cc:
```diff
@@ -0,0 +1,154 @@
+  , _connections(c)
+  , _local_service(s) {}
+
+ss::future<cluster::errc> client::produce(
+  model::topic_partition tp, ss::chunked_fifo<model::record_batch> batches) {
+    vlog(log.trace, "producing {} batches to {}", batches.size(), tp);
+    auto leader = _leaders->local().get_leader(
+      model::topic_namespace_view(model::kafka_namespace, tp.topic),
+      tp.partition);
+    if (!leader) {
+        co_return cluster::errc::not_leader;
+    }
+    produce_request req;
+    req.topic_data.emplace_back(std::move(tp), std::move(batches));
+    req.timeout = timeout;
+    auto reply = co_await (
+      *leader == _self ? do_local_produce(std::move(req))
+                       : do_remote_produce(*leader, std::move(req)));
+    vassert(
+      reply.results.size() == 1,
+      "expected a single result: {}",
+      reply.results.size());
+
+    co_return reply.results.front().err;
+}
```
question: would it be worthwhile to attempt a retry loop in here in case partition leadership changes while message is in flight?

michael-redpanda replied:
Really comes down to how similar to a kafka client we want this to be

rockwotj replied:
Right now I'm aiming for as simple as possible and I'll let the upstream processing and error handling retry, but I certainly will be adding some retry/batching logic in the client.

rockwotj replied:
So yes it's worthwhile, I'll probably put up another PR for that later.

michael-redpanda replied:
> So yes it's worthwhile, I'll probably put up another PR for that later.

Or wait for the next service to do it when they need it (*cough*)

rockwotj commented:
> Any reason this has to live in transform? Can it be usable by other 'services' (e.g. a service that may need to produce a log to a topic)

Totally - it could be used for audit log 😉 

Force push: fix commit messages, and fix the bug you found with redeclaring batches

michael-redpanda commented:
> > Any reason this has to live in transform? Can it be usable by other 'services' (e.g. a service that may need to produce a log to a topic)
> 
> Totally - it could be used for audit log 😉
> 
> Force push: fix commit messages, and fix the bug you found with redeclaring batches

What made you think I was talking about audit log 😉

michael-redpanda reviewed (approved):
lgtm but I would want another set of eyes on it that may catch any minutiae  I missed

michael-redpanda commented:
Are there any unit tests we could write for this?  Have you done any testing with it?

rockwotj commented:
I would like to write unit tests, but not quite sure how. I have tested another version of this in the `transform-dev` branch.

rockwotj commented:
Unsure how without spinning up all of Redpanda. I will look more into this tomorrow

rockwotj commented:
Took a quick peek and the options are to either wrap all the external system state I need (partition manager, shard_table, metadata_cache, etc) or to just spin up a redpanda thread fixture...

I'm leaning towards wrapping the other system functionality. Would love any input on the options.

rockwotj commented:
Okay I wrote some unit tests!

emaxerrno commented:
@rockwotj curious how this may potentially break all metrics. 

if a user writes a wasm projection and it writes to a topic downstream, wouldn't you expect metrics (existing) to increase proportionally? 

curious why the partition as the main abstraction didn't work. rfc link si good too. trying to learn.

dotnwat reviewed (commented):
> The justifications for using internal RPCs instead of the Kafka API:

Downsides to using the internal RPC:

* Circumvents any shared logic in `produce` handler that applies to all Kafka topics. A quick browse on the handler shows things like:
    * idempotency and transaction logic
    * built-in support for acks levels
    * data usage accounting
    * multi-stage producing
    * crc checks
    * null record rejection
    * schema validation
    * validation of timestamps

I think that audit logging is also considering the use of an internal RPC for producing.

dotnwat commented on src/v/kafka/server/partition_proxy.cc:
```diff
@@ -29,5 +29,13 @@ std::optional<partition_proxy> make_partition_proxy(
     }
     return std::nullopt;
 }
+std::optional<partition_proxy> make_partition_proxy(
```
> kafka/partition_proxy: support ntp creation

why do we need this?

rockwotj replied:
Sorry this should be `kafka/partition_proxy: add replicate methods` I'll fix

dotnwat replied:
it's probably the same question, tho. why is `ktp` not enough?

rockwotj replied:
I think we can stick with `ktp` as the produce path assumes this.

I think I was putting the cart in front of the horse and thinking about supporting internal namespaces (like `redpanda`) to use for wasm binaries so nothing is exposed in a public interface, but I will remove that for now as the rpc doesn't support that at the moment anyway.

dotnwat commented on src/v/kafka/server/replicated_partition.cc:
```diff
@@ -302,6 +305,13 @@ raft::replicate_stages replicated_partition::replicate(
   model::record_batch_reader&& rdr,
   raft::replicate_options opts) {
     using ret_t = result<raft::replicate_result>;
+    if (_partition->is_read_replica_mode_enabled()) {
```
makes sense to me. presumably we already protect against writing into read replicas somehow? cc @bharathv @Lazin

rockwotj replied:
FWIW we do this in the produce path directly: https://github.com/redpanda-data/redpanda/blob/bd9292c0825b2f7ab1551761d5b89ecb28d8667e/src/v/kafka/server/handlers/produce.cc#L415-L421

But this partition proxy that @mmaslankaprv introduced did a bunch of read replica checks at this level, so it seemed like the right place to do these checks, I guess doing this check here means the produce path one can be removed because it's redundant. Thoughts on where the right place is?

Related is the discussion towards the generic data path code.

dotnwat replied:
yeh moving it down into the partition layer makes sense.

dotnwat commented on src/v/transform/rpc/deps.h:
```diff
@@ -0,0 +1,123 @@
+/**
+ * Handles routing for shard local partitions.
+ */
+class partition_manager {
```
should we build interface types outside of transform::rpc so other subsystems don't need to repeat this?

rockwotj replied:
Yes I think so, or we should have a single interface type for the entire data path that the different entry points come into (kafka, internal, future protocol X). I would love to talk through the right place organizationally for this with you. I don't like how large the cluster target is and maybe we just need a "data path" target shared by this and the kafka server?

dotnwat replied:
sgtm. we can prototype in transform::* and then move it out later

dotnwat reviewed (commented):
lgtm

rockwotj commented on src/v/rpc/rpc_compiler.py:
```diff
@@ -56,12 +56,12 @@
 
 template<typename Codec>
-class {{service_name}}_service_base : public rpc::service {
+class {{service_name}}_service_base : public ::rpc::service {
```
well I have an RPC service that is wrapped in the namespace `transform::rpc`. So now anything qualified with `rpc` will lookup `transform::rpc::*`.

We have to either:
1) choose a different name for the namespace instead of `transform::rpc`. I am open to different names.
2) do this and fully qualify the name here.

I don't see another option unless I'm missing something? We've never needed this before because nobody else has tried to use an `rpc` in a nested namespace. I'm probably being overly picky, but I liked the classes introduced in this PR being namespaced as `transform::rpc::client` and `transform::rpc::service`. But I can drop this commit and use a different namespace like `data`, `io` or I guess not nest the namespace. My instinct is that each directory is a CMake target and it's own namespace as well, but I could keep it in the transform namespace (or choose another one).

dotnwat replied:
>We've never needed this before because nobody else has tried to use an rpc in a nested namespace

makes sense. i think i missed this originally

rockwotj commented:
> Downsides to using the internal RPC:
> 
>     * Circumvents any shared logic in `produce` handler that applies to all Kafka topics.

I am happy to abandon this, but there was quite a performance boost using this over the internal kafka client. I had timeboxed a day for improving the internal client (allowing pipelining of requests was the lowest hanging fruit I saw) and I ran out of time. Standing up this client (minus the work for writing the unit tests in this PR) was about 1/2 a day.

There is certainly more work in produce path w.r.t. schema validation and crc checks for certain.

The downsides of using the internal kafka client:
* It avoids a number of authz issues with the ephemeral credential, here's a series or recent work that caused me to be weary of that: https://github.com/redpanda-data/redpanda/pull/11563, https://github.com/redpanda-data/core-internal/issues/684
> if a user writes a wasm projection and it writes to a topic downstream, wouldn't you expect metrics (existing) to increase proportionally?

Personally no, because I wouldn't expect an internal cluster transform to increment a stat called redpanda_kafka_{request,response}_bytes_total. With this we approach we can have a dedicated set of stats for users that they can monitor separate (if they'd like) or combine with a sum (if they'd like). Product was aligned with this when we had the conversation a while back. But I understand the concern about updating dashboards, but I feel like it's a better solution long term. Happy to stand corrected.

> curious why the partition as the main abstraction didn't work.

Sorry I don't understand this @emaxerrno - why isn't the partition the main abstraction?

rockwotj commented on src/v/kafka/server/partition_proxy.h:
```diff
@@ -136,6 +146,19 @@ class partition_proxy {
+      model::record_batch_reader&& r,
+      raft::replicate_options opts) {
+        return _impl->replicate(
+          bi, std::forward<model::record_batch_reader>(r), opts);
```
yes I did - thank you.

dotnwat commented:
> The downsides of using the internal kafka client:
> It avoids a number of authz issues with the ephemeral credential,

i get there are issues with the internal kafka client, but we built an entirely new data path instead of dealing with those problems.

> but there was quite a performance boost using this over the internal kafka client. 

same thing.

> I had timeboxed a day for improving the internal client

then this all makes sense, thanks for the context i didn't realize it was scheduled that tightly.

rockwotj commented on src/v/transform/rpc/client.cc:
```diff
@@ -0,0 +1,154 @@
+                    timeout,
+                    [&req](impl::transform_rpc_client_protocol proto) mutable {
+                        return proto.produce(
+                          req.share(),
```
No because `req` is created only in the caller of this method.

dotnwat replied:
>No because req is created only in the caller of this method.

then why share at all instead of moving it?

rockwotj replied:
I use it below in the error case to simplify the types - but maybe the return type needs wrapped in a result type?

In the future we'd want this if we do retries at this level (suggested by Mike) but I can remove the functionality until we actually need it.

michael-redpanda replied:
I thought the response contains a list of topics and their respective result code?  It looks like you're expecting the topic response list to be in the same order as the topic request list.

michael-redpanda replied:
Sorry I'm a little confused by this.  The `produce_request` may contain one or more NTPs to produce data to.  Shouldn't the response be a list of NTPs with their respective error code?  Why is it the same error code for each NTP?  Am I misreading this?

rockwotj replied:
So by default if the RPC succeeds we get a list of ntp's and their respective error codes, there is no ordering guarantees between the request and response.

If the rpc **fails** then we will respond with that error code for all the ntps so error handling can be uniform.

michael-redpanda replied:
Oh, gotcha this is a situation where the rpc 'mechanism' runs into an issue.  Thanks

rockwotj replied:
Yes. An example of this is the timeout of the network response.

rockwotj commented on src/v/model/record_batch_reader.cc:
```diff
@@ -202,17 +203,18 @@ record_batch_reader make_fragmented_memory_record_batch_reader(
-      std::min(batches.elements_per_fragment(), batches.size()));
-    for (size_t i = 0; i < batches.size(); ++i) {
-        if (!data_chunk.empty() && i % batches.elements_per_fragment() == 0) {
+    data_chunk.reserve(std::min(elements_per_fragment, batches.size()));
```
Nice catch!

rockwotj commented on src/v/model/record_batch_reader.cc:
```diff
@@ -202,17 +203,18 @@ record_batch_reader make_fragmented_memory_record_batch_reader(
 std::vector<record_batch_reader::storage_t>
-make_storage_batches_from_fragmented_vector(
-  fragmented_vector<model::record_batch> batches) {
+make_storage_batches_from_fragmented_container(Container batches) {
```
Good point! Fixed the name.

emaxerrno commented:
> > The downsides of using the internal kafka client:
> > It avoids a number of authz issues with the ephemeral credential,
> 
> i get there are issues with the internal kafka client, but we built an entirely new data path instead of dealing with those problems.
> 
> > but there was quite a performance boost using this over the internal kafka client.
> 
> same thing.
> 
> > I had timeboxed a day for improving the internal client
> 
> then this all makes sense, thanks for the context i didn't realize it was scheduled that tightly.

@rockwotj  - agreeing w/ noah. adding an entire new data api seems like a big decision to add. i.e.: wonder of additional security implications in the future, etc. 

w.r.t metrics, users expect that anything that produces to a topic to increase the counter of produced bytes to that topic, so you can explain it. Imagine a large organization billing use of cluster to sub organizations, so if metrics don't increase proportionally, then it's free traffic (clearly false from a usage of the machine perspective). 

My reaction - and @dotnwat and @mmaslankaprv  would know better - is that an entire new data api has a lot of risk, operationally speaking. what noah pointed out seems like good risks to think through. 

Functionally ... it also means functionality will never be avail, like transactions for downstream multi partition writes.

dotnwat commented:
> @rockwotj - agreeing w/ noah. adding an entire new data api seems like a big decision to add. i.e.: wonder of additional security implications in the future, etc.

i suspect the security implications are improved because we are as locked down as internal rpcs. for example, it's the same model we have with ephemeral users and how panda proxy / schema registry are locked down.

> My reaction - and @dotnwat and @mmaslankaprv would know better - is that an entire new data api has a lot of risk, operationally speaking. what noah pointed out seems like good risks to think through.
>Functionally ... it also means functionality will never be avail, like transactions for downstream multi partition writes.

if we are talking about O(days) costs (which it seems to be) between kafka client improvements and using internal rpc usage then I think we can use delivery date as a tie breaker because abandoning the decision later wouldn't be terribly expensive.

> is that an entire new data api has a lot of risk

we had it for `coproc` and we have it currently for TS read path. it's not ideal, but the alternative may cost more time than we are willing to spend.

rockwotj commented:
> w.r.t metrics, users expect that anything that produces to a topic to increase the counter of produced bytes to that topic, so you can explain it. Imagine a large organization billing use of cluster to sub organizations, so if metrics don't increase proportionally, then it's free traffic (clearly false from a usage of the machine perspective).

Just want to comment here. From my reading I don't think we emit these metrics on anything but a per api_key breakdown, so you can't account by client or by topic as far as I know.

Also, this also means we're prevented from optimizations like skipping the network because of these metrics, which feels broken. If an org wants to provide redpanda as a service internally and add Wasm there is a lot of other things to consider, one is they may want to bill wasm **differently** than "normal" traffic. This means that having a separate metric and giving customers the choice seems better. I believe product is aligned with this choice as well

mattschumpert commented:
@rockwotj , @dotnwat , @emaxerrno ,

The spirit of the objections is completely valid but in practice only a couple of them hold as legitimate concerns, tipping the balance to RPC if we can eliminate the important concerns.

**Points For RPC**

+ we've proven at this point the internal RPC is much much much more efficient (tput)
+ its _not_ paving much of a new/risky data patch since RPC its plenty used for other subsystems as a data path, and months of testing has/will de-risk using it
+ It's not a factor from a security perspective (RPC is already in the data path, has TLS, etc and is well understood by admins)
+ cost to switch back delays this delivery a nontrivial amount 
+ we actually don't want to count WASM in metrics for things like 'active connections' (means external connections) nor 'count' them in our capacity clamps (config clamps, client quotas) for max connections from a particular client etc.  WASM should have full reign over the partitions  its emitting to without being 'counted' against those IMO

**Points neutral-ish**

+ Metrics around transforms execution & lifecycle itself are clearly new and distinct, as has been validated with users.
+ Metrics counting bytes into a topic being as a proxy for chargeback is a very suspect use case because the traffic is internal (if you want to chargeback by ingress/egress you'll almost certainly do it at the front door of the cluster and if you do consider the transform as 'belonging' to the BU being charged you want to charge on compute, not internal traffic. Minimal to no factor here.   The overall message counts however is a legitimate concern (below)

**Points potentially for Kafka API?**

Important

+ Schema validation (this must occur on the output topic as configured since this is a topic-wide policy irrespective of source)
+ Other metrics: We need message counts and other topic/partition metrics relating to whole messages (application-centric) to _remain_ accurate on output topics, in contrast to system-centric things like bytes/connections focused on external traffic.

Nice-to-have/Smaller

+ Null record rejection (this could happen in a transform and maybe help us from a perf perspective)?
+ Its' probably still important as some stage to measure byte throughput of the transform subsystem overall for supportability (troubleshooting/perf management).  We don't get this for free anymore. But we can probably live without it for a while

rockwotj commented:
> Schema validation (this must occur on the output topic as configured since this is a topic-wide policy irrespective of source)
> Null record rejection (this could happen in a transform and maybe help us from a perf perspective)?
This does happen in the transform, but I will also add these to the internal RPC path. I think overtime we can reuse a lot of the code between the two paths (that's how I'm familiar with doing this in past systems), by having an a subsystem that is "the abstract interface of truncationality that Redpanda exposes" and kafka and transforms both consuming from that.

> Its' probably still important as some stage to measure byte throughput of the transform subsystem overall for supportability (troubleshooting/perf management). We don't get this for free anymore. But we can probably live without it for a while

FWIW I already have a stat for this in transform-dev per function

dotnwat commented:
> The spirit of the objections is completely valid but in practice only a couple of them hold as legitimate concerns, tipping the balance solidly to RPC esp if we can eliminate the 'important' concerns.

@mattschumpert thanks so much for the detailed reasoning here. i agree with all of it, and having this context is really nice to justify the technical aspects of the decision to move forward with rpc.

rockwotj commented on src/v/transform/rpc/service.cc:
```diff
@@ -0,0 +1,146 @@
+  ss::chunked_fifo<transformed_topic_data> topic_data,
+  model::timeout_clock::duration timeout) {
+    ss::chunked_fifo<transformed_topic_data_result> results;
+    co_await ss::parallel_for_each(
```
Good point will limit concurrency here (what do we do for kafka?)

Ordering: no promises outside of partitions as usual semantics. Within a partition my answer is whatever the raft layer does with a reader :) Which is general we'll apply them in order. Do you have further questions here?

michael-redpanda replied:
We good with a semaphore value of 10?

Any other questions irt ordering?

This all looks good to me, but I'm not sure if there's some trick wrt ordering I'm missing

rockwotj replied:
> We good with a semaphore value of 10?

For now yes? We can tweak it, make it a config or adaptive as we do some testing.

This RPC does not make any assurances of ordering of the topic partition produce requests see: https://github.com/redpanda-data/redpanda/pull/13277#discussion_r1340375091

rockwotj commented:
Force push: limit concurrency in produce (hard coded now, maybe something fancier later).

dotnwat commented:
i guess that force was to fix formatting?

rockwotj commented:
Sorry the force push was to fix a compilation error (missing closing the lambda)

dotnwat commented on src/v/model/record_batch_reader.cc:
```diff
@@ -224,6 +224,8 @@ make_fragmented_memory_storage_batches(Container batches) {
                 data.emplace_back(std::exchange(data_chunk, {}));
             }
+            data_chunk.reserve(
+              std::min(elements_per_fragment, i - batches.size()));
```
isn't `i - batches.size()` always negative so it's close to size_t::max?

rockwotj replied:
Good catch!

rockwotj commented on src/v/transform/rpc/service.cc:
```diff
@@ -0,0 +1,153 @@
+    // TODO: schema validation
+    auto ec = co_await _partition_manager->invoke_on(
+      *shard,
+      [&ktp, timeout](
```
Done.

rockwotj commented on src/v/kafka/server/replicated_partition.h:
```diff
@@ -157,12 +157,12 @@ class replicated_partition final : public kafka::partition_proxy::impl {
 
     ss::future<result<model::offset>>
-      replicate(model::record_batch_reader, raft::replicate_options);
+      replicate(model::record_batch_reader, raft::replicate_options) final;
```
Correct! The produce path doesn't use the interface but does use `replicated_partition`.

rockwotj commented:
Force push: https://github.com/redpanda-data/redpanda/pull/13277#discussion_r1327904281 and https://github.com/redpanda-data/redpanda/pull/13277#discussion_r1327891888

dotnwat reviewed (dismissed):
lgtm

rockwotj commented:
Force push: rebased on top of dev and fixed the tests after #12906 via https://github.com/redpanda-data/redpanda/pull/13277/commits/d4ba251a2529bb9c42e09a9057c12c8046efedf7

rockwotj commented:
Force push: fixed issue with incorrect status code being returned when we moved the read replica check down a level

rockwotj commented:
CI Failure: https://github.com/redpanda-data/redpanda/issues/13181
